### PR TITLE
Split filesystem impl, fixes https://github.com/mozilla/thimble.webmaker.org/issues/1150

### PR DESCRIPTION
**Pull request by humphd**
This was an attempt to move all the Blob URL caching into `BracketsFiler`, but it turns out to be impossible.  I've made a bunch of good changes though, so I want to take the rest of it.  A few of the things you'll find in it:
* separation of Blob URL preloading into its own module (`FileSystemCache`)
* removal of async versions of URL lookup
* decoding of paths if you use `BracketsFiler` directly vs. via Brackets API
* optional file consistency check on writes, so we don't hit https://github.com/mozilla/thimble.webmaker.org/issues/1150
* I've made the rewriters to be properly async with `setTimeout` to yield control to the main thread more often.  Since almost none of the code is async, a really big web page could block the main thread for a long time.  This will do it in chunks.

### CODE REVIEW

File: src/extensions/default/bramble/main.js [hunk 1]
```diff
@@ -142,7 +143,7 @@ define(function (require, exports, module) {
 
             deferred.always(function() {
                 // Preload BlobURLs for all assets in the filesystem
-                BlobUtils.preload(root, function(err) {
+                FileSystemCache.refresh(function(err) {
```
Comment (gideonthomas) [inline src/extensions/default/bramble/main.js]: This is awesome! I no longer need to figure out a root to pass in!

File: src/filesystem/impls/filer/FileSystemCache.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    // Walk the project root dir and make sure we have Blob URLs generated for
+    // all file paths.
+    exports.refresh = function(callback) {
+        var fs = BracketsFiler.fs();
+
+        function _getUrlAsync(filename, callback) {
+            var decodedFilename = decodePath(filename);
+            var cachedUrl = BlobUtils.getUrl(filename);
+            if(cachedUrl && cachedUrl !== filename) {
```
Comment (gideonthomas) [inline src/filesystem/impls/filer/FileSystemCache.js]: do you need the "if cachedUrl exists" check? Shouldn't the `cachedUrl !== filename` check suffice?

File: src/filesystem/impls/filer/FileSystemCache.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                setTimeout(function() {
```
Comment (gideonthomas) [inline src/filesystem/impls/filer/FileSystemCache.js]: Can you add a comment above for why we do this?

File: src/filesystem/impls/filer/FileSystemCache.js (hunk 2 above)
Comment (humphd) [inline src/filesystem/impls/filer/FileSystemCache.js]: Yeah, you're right.  This is a hold-over from the old way of doing it.  I can remove this in a few places.

File: src/filesystem/impls/filer/FilerFileSystem.js [hunk 4]
```diff
@@ -316,7 +322,9 @@ define(function (require, exports, module) {
                 return;
             }
 
-            if (options.hasOwnProperty("expectedHash") && options.expectedHash !== stats._hash) {
+            if (_doConsistencyCheck                    &&
```
Comment (gideonthomas) [inline src/filesystem/impls/filer/FilerFileSystem.js]: This will fix the "External Changes detected" thing right?

Comment (humphd) [inline src/filesystem/impls/filer/FilerFileSystem.js]: Correct.  We just won't bother checking the hash/cached version against this content, we'll just write it out to disk no matter what.

Comment (gideonthomas) [inline src/filesystem/impls/filer/FilerFileSystem.js]: Neat!

File: src/filesystem/impls/filer/lib/transforms.js [hunk 5]
```diff
@@ -11,7 +11,7 @@ define(function (require, exports, module) {
     var ExtensionUtils = require("utils/ExtensionUtils");
     var marked = require("thirdparty/marked/marked.min");
     var StartupState = require("bramble/StartupState");
-    var Path = BracketsFiler.Path;
+    var Path = require("filesystem/impls/filer/FilerUtils").Path;
```
Comment (gideonthomas) [inline src/filesystem/impls/filer/lib/transforms.js]: Why this change?

**Comment by gideonthomas** [comment]
a few questions, r=me with that!

File: src/filesystem/impls/filer/lib/transforms.js (hunk 5 above)
Comment (humphd) [inline src/filesystem/impls/filer/lib/transforms.js]: Yeah, it was from a failed rewrite attempt to remove `BracketsFiler`.  I'll revert.
